opts.3
.TH "opts.js" "3"
.PP
Find the full documentation, source code, and examples at \fIhttps://khtdr.com/opts\fP

.SH "NAME"
.PP
opts.js - a command line parser for options and arguments

.SH "VERSION"
.PP
v2.0.1

.SH "SYNOPSIS"
.PP
The following example uses a custom \fBversion\fP function, and opts in to the automatic help text. No pun intended.

.RS
.nf
\fCvar opts = require('opts');

var options = [
  { short       : 'v'
  , long        : 'version'
  , description : 'Show version and exit'
  , callback    : function () { console.log('v1.0'); process.exit(1); }
  }
];

opts.parse(options, true);
console.log('Example 1');
process.exit(0);
\fP
.fi
.RE
.PP
See \fIhttps://raw.githubusercontent.com/khtdr/opts/master/examples/example1.js\fP

.SS "running:"
.RS
.nf
\fC$ node ./example1
\fP
.fi
.RE
.SS "produces:"
.RS
.nf
\fCExample 1
\fP
.fi
.RE

.SS "running:"
.RS
.nf
\fC$ node ./example1 --help
\fP
.fi
.RE
.SS "produces:"
.RS
.nf
\fCUsage: node ./example1 [options]
Show this help message
   --help
Show version and exit
   -v, --version
\fP
.fi
.RE

.SS "running:"
.RS
.nf
\fCnode ./example1 -v
\fP
.fi
.RE
.SS "produces:"
.RS
.nf
\fCv1.0
\fP
.fi
.RE



.SH "INSTALLATION"
.PP
You do not need to use NPM or any package manager. It is written in plain-old Javascript and can be downloaded and included in your Node.js project, as-is. All of the examples use this approach.
.RE
See \fIhttps://github.com/khtdr/opts/tree/master/examples\fP

.SS "Stand-alone version"
.RS
.nf
\fCcd /path/to/your/project
curl -o opts.js https://raw.githubusercontent.com/khtdr/opts/master/src/opts.js
\fP
.fi
.RE

.SS "NPM version"
.RS
.nf
\fCnpm install opts
\fP
.fi
.RE

.SH "USAGE"
.SS "LOADING"
.PP
With classic syntax:
.RS
.nf
\fCvar opts = require('opts');
opts.parse(options, arguments, help);
\fP
.fi
.RE

.PP
With modern syntax:
.RS
.nf
\fCimport * as opts from 'opts';
opts.parse(options, arguments, help);
\fP
.fi
.RE

.PP
If you installed \fCopts\fP with NPM, the typescript definitions should automatically be available in your editor. Otherwise you can download the .d.ts file manually.
.RE
See \fIhttps://raw.githubusercontent.com/khtdr/opts/master/src/opts.d.ts\fP

.SS "CONFIGURING"
.PP
\fIopts.parse(options, arguments, help)\fP

.PP
Options are flag-arguments. Arguments are everything else. Consider the following hypothetical command for starting a server that listens on \fIhttp://0.0.0.0:4000\fP

.RS
.nf
\fCnode ./my-app start --host 0.0.0.0 -p 4000
\fP
.fi
.RE

.PP
In this example, the options are \fI\-\-host 0.0.0.0\fP and \fI\-p 4000\fP. The argument is \fIstart\fP. The arguments can be after, before, or among the options.

.SS "options"
.PP
\fCoptions\fP is an array of option objects. Each option in the array can have the following fields. None are required, but you should at least provide a short or long name.

.RS
.nf
\fClet options = [
  { short       : 'l',
    long        : 'list',
    description : 'Show a list',
    value       : false, // default false
    required    : true, // default false
    callback    : function (value) { ... },
  }, // ... followed by more options
];
\fP
.fi
.RE

.SS "arguments"
.PP
\fCarguments\fP require less configuration. This is an optional argument to \fCopts.parse\fP:

.RS
.nf
\fClet arguments =
  { name     : 'script',
    required : true, // not required by default
    callback : function (value) { ... },
  };
\fP
.fi
.RE

.SS "help text generator"
.PP
Finally, you can add an automatically generated help message by passing
a last parameter of \fItrue\fP. This is also an optional argument to \fCopts.parse\fP.

.RS
.nf
\fCopts.parse(options, true);
// or if you want more control, you can do:
/*
  options.push({
    long        : 'help',
    description : 'Show this help message',
    callback    : require('opts').help,
  }
  opts.parse(options);
*/
\fP
.fi
.RE



.SH "AUTHOR / CHANGELOG / LICENSE"
.PP
Email: ohkay@khtdr.com

.PP
Relatively unchanged since 2010.
.RE
See \fIhttps://github.com/khtdr/opts/blob/master/CHANGES.org\fP

.PP
BSD 2-Clause License
.RE
See \fIhttps://github.com/khtdr/opts/blob/master/LICENSE\fP


CHANGES.org
* Changelog

** 2010 =v1.0.0=
initial commit

** 2012 =v1.2.0=
~sys.puts~ -> ~console.log~

** 2012 =v1.2.1=
added version number to exports

** 2012 =v1.2.2=
bug fix on missing variable

** 2012 =v1.2.3=
added support for multiple params, *bad idea, reverted*

** 2017 =v1.2.4= through =v1.2.6=
examples, docs

** Oct 29, 2019 =v1.2.7=
no code changes, added typescript definitions

** Jun 29, 2020 =v2.0.0=
- Added ~values()~ to retrieve all the flag values at once.
#+BEGIN_SRC javascript
opts.parse([{
    value: true, required: true,
    short: 'n', long: 'num'
}, {
    short: 'b', long: 'bool'
}])
const { n, b } = opts.values()
#+END_SRC

- Changes the output of the auto-generated help text.
    1) Splits and pops `argv[0]`
    2) Replaces `pwd` with "." in `argv[1]`

#+BEGIN_SRC bash
ok@laptop: opts [master]× » node examples/example2.js --help
Usage: /home/ok/.nvm/versions/node/v12/bin/node /home/ok/khtdr.com/opts/examples/example2.js [options]
Show this help message
    --help
    ...
#+END_SRC

Becomes:
#+BEGIN_SRC bash
ok@laptop: opts [master]× » node examples/example2.js --help
Usage: node ./examples/example2.js [options]
Show this help message
    --help
    ...
#+END_SRC

This is a *breaking change* if you are parsing the help text.


** Aug 14, 2020 =v2.0.1=
- Bumped the exported version number, having forgotten to in the previous release. NPM requires a new version for a new release.
- Moved to org-mode documentation, added a man page
  #+BEGIN_SRC bash
  curl -o opts.3 https://raw.githubusercontent.com/khtdr/opts/master/opts.3
  man ./opts.3
  #+END_SRC


src/opts.js
/***************************************************************************
Author   : Joey Mazzarelli
Email    : khtdr.com@gmail.com
Homepage : http://khtdr.com/opts
Source   : http://github.com/khtdr/opts.git
License  : Simplified BSD License

Copyright 2010 Joey Mazzarelli. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.

   2. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY JOEY MAZZARELLI 'AS IS' AND ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
EVENT SHALL JOEY MAZZARELLI OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those
of the authors and should not be interpreted as representing official policies,
either expressed or implied, of Joey Mazzarelli.
***************************************************************************/

var puts        = console.log
  , values      = {}
  , args        = {}
  , argv        = []
  , errors      = []
  , descriptors = {opts:[], args:[]};

/**
 * Get some version info out of it
 */
exports.version = '2.0.1';

/**
 * Add a set of option descriptors, not yet ready to be parsed.
 * See exports.parse for description of options object
 *
 * Additionally, it takes a namespace as an argument, useful for
 * building options for a library in addition to the main app.
 */
exports.add = function (options, namespace) {
  for (var i=0; i<options.length; i++) {
    options[i].namespace = namespace;
    descriptors.opts.push(options[i]);
  }
};

/**
 * Parse the command line options
 * @param array options  Options to parse
 * @param array args     Arguments to parse
 * @param bool  help     Automatically generate help message, default false
 *
 * ===== Options Docs =====
 * Each option in the array can have the following fields. None are required, 
 * but you should at least provide a short or long name.
 *   {
 *     short       : 'l',
 *     long        : 'list',
 *     description : 'Show a list',
 *     value       : false,  // default false
 *     required    : true,   // default false
 *     callback    : function (value) { ... },
 *   }
 *
 * You can add an automatically generated help message by passing
 * a second parameter of <true> or by including the option;
 *   {
 *     long        : 'help',
 *     description : 'Show this help message',
 *     callback    : require('./opts').help,
 *   }
 *
 * ===== Arguments Docs =====
 * Arguments are different than options, and simpler. They typically come 
 * after the options, but the library really doesn't care. Each argument
 * can have the form of:
 *   {
 *     name     : 'script',
 *     required : true,      // default false
 *     callback : function (value) { ... },
 *   }
 */
exports.parse = function (options, params, help) {

  if (params === true) {
    help = true;
    params = [];
  } else if (!params) {
    params = [];
  } else {
    for (var i=0; i<params.length; i++) {
      descriptors.args.push(params[i]);
    }
  }

  if (help) {
    options.push({ long        : 'help'
                 , description : 'Show this help message'
                 , callback    : exports.help
                 });
  }
  for (var i=0; i<options.length; i++) {
    descriptors.opts.unshift(options[i]);
  }
  options = descriptors.opts;

  var checkDup = function (opt, type) {
    var prefix = (type == 'short')? '-': '--';
    var name = opt[type];
    if (!opts[prefix + name]) {
      opts[prefix + name] = opt;
    } else {
      if (opt.namespace && !opts[prefix + opt.namespace + '.' + name]) {
        opts[prefix + opt.namespace + '.' + name] = opt;
        for (var i=0; i<descriptors.opts.length; i++) {
          var desc = descriptors.opts[i];
          if (desc.namespace == opt.namespace) {
            if (type == 'long' && desc.long == opt.long) {
                descriptors.opts[i].long = opt.namespace + '.' + opt.long;
            } else if (type == 'short') {
              delete descriptors.opts[i].short;
            }
          }
        }
      } else {
        puts('Conflicting flags: ' + prefix + name + '\n');
        puts(helpString());
        process.exit(1);
      }
    }
  };

  var opts = {};
  for (var i=0; i<options.length; i++) {
    if (options[i].short) checkDup(options[i], 'short');
    if (options[i].long) checkDup(options[i], 'long');
  }

  for (var i=2; i<process.argv.length; i++) {
    var inp = process.argv[i];
    if (opts[inp]) {
      // found a match, process it.
      var opt = opts[inp];
      if (!opt.value) {
        if (opt.callback) opt.callback(true);
        if (opt.short) values[opt.short] = true;
        if (opt.long) values[opt.long] = true;
      } else {
        var next = process.argv[i+1];
        if (!next || opts[next]) {
          var flag = opt.short || opt.long;
          errors.push('Missing value for option: ' + flag);
          if (opt.short) values[opt.short] = true;
          if (opt.long) values[opt.long] = true;
        } else {
          if (opt.callback) opt.callback(next);
          if (opt.short) values[opt.short] = next;
          if (opt.long) values[opt.long] = next;
          i++;
        }
      }
    } else {
      // No match. If it starts with a dash, show an error. Otherwise
      // add it to the extra params.
      if (inp[0] == '-') {
        puts('Unknown option: ' + inp);
        if (opts['--help']) puts('Try --help');
        process.exit(1);
      } else {
        argv.push(inp);
        var arg = params.shift();
        if (arg) {
          args[arg.name] = inp;
          if (arg.callback) arg.callback(inp);
        }
      }
    }
  }
  for (var i=0; i<options.length; i++) {
    var flag = options[i].short || options[i].long;
    if (options[i].required && !exports.get(flag)) {
      errors.push('Missing required option: ' + flag);
    }
  }
  for (var i=0; i<params.length; i++) {
    if (params[i].required && !args[params[i].name]) {
      errors.push('Missing required argument: ' + params[i].name);
    }
  }
  if (errors.length) {
    for (var i=0; i<errors.length; i++) puts(errors[i]);
    puts('\n' + helpString());
    process.exit(1);
  }
};

/**
 * Get the value of an option. Can be the short or long option
 * @return string|true|undefined
 */
exports.get = function (opt) {
  return values[opt] || values['-' + opt] || values['--' + opt];
};

/**
 * Get all of the values after parsing. Can be short or long.
 * @return {[flag string] string|true}
 */
exports.values = function () {
  return Object.keys(values).reduce(
    function(dict, name) {
      name = name.replace('/^-+/', '');
      dict[name] = exports.get(name);
      return dict;
    }, {}
  );
};

/**
 * Get unknown args. Could have special meaning to client
 */
exports.args = function () {
  return argv;
};

/**
 * Get an arg by name.
 * This only works if arg names were passed into the parse function.
 * @param string name Name of arg
 * @return string Value of arg
 */
exports.arg = function (name) {
  //puts(require('sys').inspect(arguments));
  return args[name];
};

/**
 * Print the help message and exit
 */
exports.help = function () {
  puts(helpString());
  process.exit(0);
};


// Create the help string
var helpString = function () {
  var exe = process.argv[0].split(require('path').sep).pop();
  var file = process.argv[1].replace(process.cwd(), '.');
  var str = 'Usage: ' + exe + ' ' + file;
  if (descriptors.opts.length) str += ' [options]';
  if (descriptors.args.length) {
    for (var i=0; i<descriptors.args.length; i++) {
      if (descriptors.args[i].required) {
        str += ' ' + descriptors.args[i].name;
      } else {
        str += ' [' + descriptors.args[i].name + ']';
      }
    }
  }
  str += '\n';
  for (var i=0; i<descriptors.opts.length; i++) {
    var opt = descriptors.opts[i];
    if (opt.description) str += (opt.description) + '\n';
    var line = '';
    if (opt.short && !opt.long) line += '-' + opt.short;
    else if (opt.long && !opt.short) line += '--' + opt.long;
    else line += '-' + opt.short + ', --' + opt.long;
    if (opt.value) line += ' <value>';
    if (opt.required) line += ' (required)';
    str += '    ' + line + '\n';
  }
  return str;
};
